server/bleep/src/webserver/studio.rs

```diff
@@ -0,0 +1,243 @@
+pub struct ListItem {
+    id: String,
+    name: String,
+    modified_at: NaiveDateTime,
+}
+
+pub async fn list(app: Extension<Application>) -> webserver::Result<Json<Vec<ListItem>>> {
+    sqlx::query_as! { ListItem, "SELECT id, name, modified_at FROM studios" }
+        .fetch_all(&*app.sql)
+        .await
+        .map_err(Error::internal)
+        .map(Json)
+}
+
+#[derive(serde::Serialize)]
+pub struct Studio {
+    name: String,
+    modified_at: NaiveDateTime,
+    context: Vec<ContextFile>,
+    messages: Vec<Message>,
+    token_counts: TokenCounts,
+}
+
+#[derive(serde::Serialize, serde::Deserialize)]
+struct ContextFile {
+    path: String,
+    hidden: bool,
+    repo: RepoRef,
+    branch: String,
+    ranges: Vec<Range<usize>>,
+}
+
+#[derive(serde::Serialize, serde::Deserialize)]
+enum Message {
+    User(String),
+    Assistant(String),
+}
+
+pub async fn get(
+    app: Extension<Application>,
+    Path(id): Path<String>,
+) -> webserver::Result<Json<Studio>> {
+    let row = sqlx::query! {
+        "SELECT id, name, context, messages, modified_at
+         FROM studios
+         WHERE id = ?",
+         id
+    }
+    .fetch_optional(&*app.sql)
+    .await
+    .map_err(Error::internal)?
+    .ok_or_else(|| Error::new(ErrorKind::NotFound, "unknown studio ID"))?;
+
+    let context: Vec<ContextFile> =
+        serde_json::from_str(&row.context).context("failed to deserialize context")?;
+    let messages: Vec<Message> =
+        serde_json::from_str(&row.messages).context("failed to deserialize message list")?;
+
+    Ok(Json(Studio {
+        modified_at: row.modified_at,
+        name: row.name,
+        token_counts: token_counts((*app).clone(), &context).await?,
+        context,
+        messages,
+    }))
+}
+
+#[derive(serde::Deserialize)]
+pub struct Patch {
+    name: Option<String>,
+    modified_at: Option<NaiveDateTime>,
+    context: Option<Vec<ContextFile>>,
+    messages: Option<Vec<Message>>,
+}
+
+pub async fn patch(
+    app: Extension<Application>,
+    Path(id): Path<String>,
+    Json(patch): Json<Patch>,
+) -> webserver::Result<Json<TokenCounts>> {
+    let mut transaction = app.sql.begin().await.map_err(Error::internal)?;
+
+    // Ensure the ID is valid first.
+    sqlx::query!("SELECT id FROM studios WHERE id = ?", id)
+        .fetch_optional(&mut transaction)
+        .await
+        .map_err(Error::internal)?
+        .ok_or_else(|| Error::new(ErrorKind::NotFound, "unknown code studio ID"))?;
+
+    if let Some(name) = patch.name {
+        sqlx::query!("UPDATE studios SET name = ? WHERE id = ?", name, id)
+            .execute(&mut transaction)
+            .await
+            .map_err(Error::internal)?;
+    }
+
+    if let Some(modified_at) = patch.modified_at {
+        sqlx::query!(
+            "UPDATE studios SET modified_at = ? WHERE id = ?",
+            modified_at,
+            id
+        )
+        .execute(&mut transaction)
+        .await
+        .map_err(Error::internal)?;
+    }
```

Comment:
we seem to set `modified_at` to current datetime on L162. do we need `patch.modified_at`?